### ISSUE TITLE
epic: Harness VLA-style agentic manipulation - retryable policy primitives, analytic staging vocabulary, harness memory, pixel-to-world grounding (arXiv:2607.08448)

### ISSUE THREAD

**Issue by yinsong1986** [strands-labs/robots#1648]
## Summary

Umbrella epic for adopting the **Harness VLA** recipe (arXiv:2607.08448, "Harness VLA: Steering Frozen VLAs into Reliable Manipulation Primitives via Memory-Guided Agents") in strands-robots. The paper wraps a **frozen** VLA in a coding-agent harness - a fixed analytic primitive vocabulary, a retryable VLA primitive with an early-return predicate, two-tier memory, and an auditable execution loop - and beats the strongest baselines by **+38.6 pts on LIBERO-Pro** and **+25.4 pts on RoboCasa365** with zero policy finetuning. Their planner backbones are literally **Codex and Claude Code**.

This is independent validation of the strands-robots thesis (LLM agent + fixed tool vocabulary + frozen policies as tools). The deltas between their harness and ours are four well-scoped capabilities, each filed as its own issue:

## Sub-issues

| # | Issue | What it adds | Paper evidence |
|---|---|---|---|
| 1 | #1644 - `stop_when=` on `run_policy` + `stopped_reason` | Turns any policy rollout into a **retryable primitive**: run until a predicate-DSL condition fires (e.g. `grasped(cube)`) or budget exhausts, and report why it stopped. All machinery (predicate registry, per-step hook, CooperativeStop) exists; this is wiring. | `vla_act`'s early-return predicate tau is the core mechanic; "retry localizes contact failures" (Key Finding 2); capped-invocation ablation shows success saturates after a few planner-selected VLA calls |
| 2 | #1645 - analytic motion primitives (`move_to`/`set_gripper`/`rotate_wrist`) | The staging/transport/release vocabulary agents need to position the robot around a policy's competence region. Backed by the existing mink IK bridges (deduplicated). | `move_to` = **61.8%** of all primitive calls on LIBERO; analytic primitives = 84.2% of calls; "analytic primitives expand the conditions under which the same frozen VLA can be reused" (Key Finding 3) |
| 3 | #1646 - `harness_memory` tool (task traces + global rules) | Persist successful solution skeletons (JSONL, re-grounded not replayed) + cross-task success rules / failure models; retrieved into agent context on later runs. | Memory worth **+56 pts** under position swaps (Goal-S 31.0% zero-shot -> 87.0% few-shot); trace = "task-level solution skeleton, not an open-loop trajectory" |
| 4 | #1647 - `get_world_point(camera, pixels)` | Deployment-shaped metric grounding: pixel picking + depth unprojection (median over samples) instead of oracle `get_body_state`. Builds on the landed `get_frame`/`get_camera_params` (#1537 PR 1). | Perception isolation: planner receives object *names*, never coordinates; localizes by indexing a depth-derived world map (Appendix E.2) |

## The composed loop (what "done" looks like)

With 1+2+4, a Strands agent can execute the paper's staged-retry pattern against any policy provider today shipping in the registry (lerobot_local, groot, pi0, ...):

```
world_pt = get_world_point("front", pixels_on_cube)        # ground (no oracle)
move_to(position=near(world_pt))                            # stage
r = run_policy(provider, instruction="pick up the red cube",
               stop_when={"predicate": "grasped", ...},     # retryable primitive
               max_steps=150)
if r.stopped_reason != "predicate":                          # inspect
    move_to(...); retry                                      # re-stage + re-invoke
move_to(above_bin); set_gripper("open")                      # analytic finish
```

With 3, the agent persists what worked and improves across sessions. Everything composes with existing surfaces: `stop_when` doubles as #1533's per-rollout success gate; recorded staged-retry rollouts flow into the same LeRobotDataset/bucket loop (#1507 blog narrative: the harness *collects better data too* - only successful contact attempts).

## Deliberately NOT borrowed

- **File-mediated REPL** (command.json / state_NN.json polling): the Strands AgentTool in-process dispatch is strictly better plumbing; we keep it.
- **LIBERO-Pro benchmark integration**: worthwhile evaluation follow-up, out of scope for this epic.
- Their stated limitations (open planner<->VLA feedback loop, no environment-reward finetuning) are places strands-robots already has stronger machinery (mesh telemetry, training providers) - no action needed.

## Suggested sequencing

#1644 (small, unblocks the pattern) -> #1647 (small) -> #1645 (medium) -> #1646 (medium, consumes the vocabulary of the first three). Each is independently shippable and reviewable.

## References

- Paper: https://arxiv.org/abs/2607.08448 (v3, 2026-07-15) - project page https://harnessvla.github.io/
- Related in-repo: #1533 (task-level pick_place provider - the scripted cousin of the agentic loop), #1537 (get_frame groundwork), #1507 (blog draft - future post material: "the agent harness that improves your frozen VLA")

**Comment by yinsong1986** [comment]
Project board routing (PAT lacks org Projects:write): please add to https://github.com/orgs/strands-labs/projects/2 with Status=Ready, Priority=P1.

**Comment by yinsong1986** [comment]
Status update on this epic: all four sub-issues have their PRs open, review feedback addressed, CI green, and re-review requested.

| Sub-issue | PR | State |
|---|---|---|
| #1644 `stop_when=` + `stopped_reason` | #1656 | Review fixes pushed (`8067a7df`): pre-rollout probe of the clause's referenced bodies/joints against the live scene (a typo'd name is now an up-front structured error instead of a clause that silently never fires and burns the budget - including on backends without body lookups), one payload shape across episode counts (`stopped_reason`/`stopped_reasons`/`steps_used` on the multi-episode aggregate), `stopped_reason="error"` json block on every error exit path, per-step guard hoisted. 10 new tests, 7 fail pre-fix. CI green, re-review requested. |
| #1645 analytic motion primitives | #1654 | Review fix pushed (`d011f4d0` + upstream merge to clear the #1643 conflict): `move_to` no longer commands the gripper - shared `_gripper_actuator_ids()` classifier, gripper excluded from IK restart seeding and the command set, channel HELD at live position every tick; grasp preservation is now a documented contract, pinned on the review's exact restart-path repro (jaw `-0.194 -> +1.107` pre-fix, held post-fix) plus a jaw assertion in the SO-100 integration journey. Follow-up filed: #1658 (registry gripper metadata). CI green, re-review requested. |
| #1646 `harness_memory` | #1651 | Review fixes pushed (`62347e6d`, CodeQL fixed earlier in `08b16b91`): UTF-8 everywhere (round-trips under `LC_ALL=C` pinned in subprocess tests), atomic `save_trace` (temp files + `os.replace`; no torn pairs, no ghost keys), full re-validation on load (`load_trace`/`list_tasks` never echo tampered or corrupt store content into planner context), read-only actions no longer mkdir. 11 new tests, 10 fail pre-fix. CI green, re-review requested. |
| #1647 `get_world_point` | #1649 | Review fixes pushed (`98497acf`): non-string `camera_name` degrades to a structured error (never a `TypeError` through the envelope), and explicit MJCF camera intrinsics are honored - `K` from `cam_intrinsic`/`cam_sensorsize`/`cam_resolution` with rasterizer-calibrated sign conventions (note: `cy = res_h/2 - principal_y/pixel_h`, minus on both axes - calibration protocol in the code), fovy fallback kept. 7 new tests fail pre-fix, incl. a non-circular depth-blob round-trip on the review's repro camera (~25 cm silent error pre-fix). CI green, re-review requested. |

With these four merged, the composed staged-retry loop in this epic's description is executable end-to-end via the tool surface (ground with `get_world_point`, stage with `move_to`, invoke `run_policy(stop_when=...)`, inspect `stopped_reason`, re-stage/retry, finish with `move_to` + `set_gripper`), and `harness_memory` persists what worked across sessions.

Board note: the PAT on this box lacks project-item scope, so #1658 (and this epic's status move to "In review") need a manual board update.

**Comment by cagataycali** [comment]
Closing as out of scope for 1.0.

The roadmap seals the supported policy, robot and embodiment list as part of theme 3, and decides the starter kit in theme 5, so holding an open issue per candidate provider, per registry expansion, per additional bridge or per hardware-in-the-loop lane pre-empts a decision that is about to be made deliberately. #469 is a good example of why the titles mislead: `KimodoPolicy` ships, but its namesake feature does not - zero files under `policies/kimodo/` mention a constraint - so the heading overstates what is left.

Re-file against the sealed 1.0 scope if it survives that decision.

Closed as part of the 0.5 -> 1.0 backlog cleanup.
